### ISSUE TITLE
8.3-Tips for GAN, WGAN

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#53]
GAN 是以不好 Train而聞名的, 所以接下來講一些 GAN 訓練的小技巧(只挑一個最知名的來講, WGAN)

### JS Divergence的問題

在講 WGAN 之前先講 JS Divergence的問題

最早的 GAN 要 Minimize 的是 JS Divergence

```
PG 跟 Pdata 有一個非常關鍵的特性是PG 跟 Pdata 它們重疊的部分往往非常少
```

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/78f0066d-a3fe-4c9f-86f5-5b2176916c00)

第一個理由是來自於 Data 本身的特性
因為你想 PG 跟 Pdata 它們都是要產生圖片的, 圖片其實是高維空間裡面的
在高維空間裡面 隨便 Sample 一個點 它通常都沒有辦法構成一個二次元人物的頭像
二次元人物的頭像它的分布啊在高維的空間中其實是非常狹窄的

第二個理由是我們是從來都不知道 PG 跟 Pdata 長什麼樣子
我們對 PG 跟 Pdata 它的分布的理解其實來自於 Sample
所以也許 PG 跟 Pdata 它們是有非常大的 Overlap 的範圍

但是我們實際上在計算它們的 Divergence 的時候我們用的是Sample的方式
從 Pdata 裡面 Sample 一些點出來 從 PG 裡面 Sample 一些點出來
如果 Sample 的點不夠多(不夠密) 去記算是這兩個 Divergence , 會很容易誤以為Divergence 很高了

JS Divergence 有個特性
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1cb08d83-6c6a-4807-ab19-85477efcccbd)

```
沒有重疊的分布 JS Divergence 算出來就永遠都是 Log2 (不管這兩個分布長什麼樣 )
```

從實際操作的角度會發現, 分辨 Real 的 Image 跟 Generated Image 正確率幾乎都是 100% 
因為你 Sample 的圖片根本就沒幾張
對你的 Discriminator 來說, 機器可以直接用硬背的 ,知道說誰是 Real 誰是 Fake, 把這兩組圖片分開

我們本來會期待說 這個 Discriminator 的 這個 Binary Classifier Loss  越來越大(這樣才可以訓練)
代表我們的 Generated Data 跟 Real 的 Data 越來越接近
但實際上你根本就沒有辦法看出你的 Generator有沒有越來越好

所以過去(尤其是在還沒有 WGAN 這樣的技術, 那個 Loss 隨著訓練的過程會慢慢變小)的時候 
Train GAN 真的就很像巫術 黑魔法, 根本就不知道你 Train 的時候有沒有越來越好
那時候做法就是用人眼守在電腦前面看
發現結果不好就咔掉, 重新用一組 Hyperparameter 重新調一下 Network

既然是 JS Divergence 的問題 , 會不會換一個換一種 Divergence就可以解決這個問題?

# WGAN
當你用 Wasserstein Distance來取代 JS Divergence 的時候, 這個 GAN 就叫做 WGAN
Wasserstein Distance(這個 W 其實這邊是唸 V 不是唸 不是發 W 的音), 解決JS Divergence 的問題方法

假設你有兩個 Distribution 
一個 Distribution 我們叫它 P
另外一個 Distribution 我們叫它 Q
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/4d8a7fe5-179b-4971-9768-2027167db70d)

Wasserstein Distance 它計算的方法就是想像你在開一台推土機
推土機的英文叫做 Earth Mover, 所以 Wasserstein Distance 又叫 Earth Mover Distance
你把 P 想成是一堆土, 把 Q 想成是你要把土堆放的目的地
Wasserstein Distance = 推土機把 P 這邊的土 挪到 Q 所移動的平均距離

但是如果是複雜的 Distribution 要算 Wasserstein Distance 就有點困難了
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/b5c5480d-f067-423d-8476-fcbd058b5ce1)

可能的 Moving Plans(把 Q 把 P 重新塑造成 Q 的方法)有無窮多種
為了讓 Wasserstein Distance 只有一個值

```
Wasserstein Distance 的定義:窮舉所有的 Moving Plans, 可以讓平均的距離最小的計畫之距離
```

所以其實要計算 Wasserstein Distance 是挺麻煩的 
居然還要解一個 Optimization 的問題才能算 Wasserstein Distance

### 計算Wasserstein Distance 的好處
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/60aa36a0-8e04-4357-9e97-dfb8e4f341a6)

如果觀察 Wasserstein Distance 的話會知道說, 從左到右我們的 Generator 越來越進步

演化的例子, 眼睛的生成
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/869da571-dabd-4f19-974f-4bfce249c98a)

那有一些生物它有非常原始的眼睛(比如說有一些細胞具備有感光的能力)
如何產生這麼複雜的人類眼睛器官呢?
中間其實是有很多連續的步驟(是極其緩慢, 每次都只改動一點)
感光的細胞可能會出現在一個比較凹陷的地方, 可以接受來自不同方向的光源
乾脆把凹陷的地方蓋起來並且放一些液體增加感光
然後最後就變成了人的眼睛 

當你使用 WGAN (Wasserstein Distance)來衡量你的 Divergence 的時候其實就製造了類似的效果
本來 PG0 跟 Pdata 非常地遙遠, 要它一步從左跑到右(讓這個 PG0 跟 Pdata直接 Align 在一起)是極其困難的

### Pdata 跟 PG 的 Wasserstein Distance計算

解下面這個 Opimilazion 的 Problem 得到的值就是 Wasserstein Distance
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/af86c5ef-b3b0-42f5-8f48-51dcf2b5fad0)

y如果是從 Pdata 來的, 那我們要計算它的 Ey 的期望值->Ey~Pdata 越大越好
y 如果是從 PG 來的, 我們計算它的 Ey 的期望值,但是前面乘上一個負號->Ey~PG 越小越好

還有一個限制是, D 不能夠是一個隨便的 Function, E 必須要是一個 1-Lipschitz 的 Function
1-Lipschitz : 足夠平滑的 Function 它不可以是變動很劇烈的 Function
為什麼足夠平滑這件事情是非常重要的呢?

在藍色的點跟綠色的點(也就是真正的 Image跟 Generated 的 Image)

如果沒有給予這個限制來訓練此GAN
Discriminator 會給 Real 的 Image 無限大的正值
給 Generated 的 Image 無限大的負值

只要這兩堆 Data 沒有重疊你算出來的值都是無限大

要求你的 Discriminator 夠平滑的時候
假設 Real Data 跟 Generated Data它們距離比較近
你就沒有辦法讓 Real 的 Data 值非常大 然後 Generated 的值非常小
Real 和 Generated 的值差異過大->Discriminator 變化劇烈->不是 1-Lipschitz

怎麼確保 Discriminator一定符合 1-Lipschitz Function 的限制?
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/3839dee3-c390-4a12-8172-0857db441811)

最早做了一個比較 Rough 比較粗糙的處理的方法
Train Network 的時候, 要求參數得在 C 跟 -C 之間
如果超過 C 用 Gradient Descent Update , 小於 -C 就直接設為 -C 

但其實這個方法但它並沒有真的去解這個Optimization 的 Problem
它並沒有真的讓 Discriminator 符合1-Lipschitz Function 的限制

接下來就有其它的想法 EX:Gradient Penalty Gradient Penalty
而現今效果最好的是 Spectral Normalization

雖然說已經有 WGAN 但其實GAN 仍然是以很難把它 Train 起來而聞名的

GAN 有一個本質上困難的地方, Discriminator 跟 Generator 它們互動的過程是自動的

Discriminator 做的事情是要分辨真的圖片跟產生出來的的圖片的差異
Generator 在做的事情它是要去產生假的圖片騙過 Discriminator
Generator 跟 Discriminator 它們是互相砥礪才能互相成長
只要其中一者停止訓練(進化)另外一個就會跟著停下訓練
只能祈禱每次 Train Discriminator 的時候它的 Loss 都是有下降的
如果有一次沒有下降那整個 Training 很有可能就會變就會慘掉

你可以在網路上找到滿坑滿谷的GAN 的 Tips 給自己參考,有沒有用不好說
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/b1728a2d-b58c-4e5e-b4ab-2becbe6e585b)
